### PR TITLE
Add test case to reproduce malformation of remote_write data during retries when sample_age_limit is applied 

### PR DESCRIPTION
**Pull request by FUSAKLA**
As discussed here https://github.com/prometheus/prometheus/pull/14078#issuecomment-2113195144 this PR adds a test case, that reproduces the issue reported in https://github.com/prometheus/prometheus/issues/13979 and fixed in https://github.com/prometheus/prometheus/pull/14078

The test case unfortunately relies on usage of `time.Now()` because of the
https://github.com/prometheus/prometheus/blob/5a218708f1f755a67b3d9b68549ccbbb6463d850/storage/remote/queue_manager.go#L634 which makes it hard to test without refactoring it.

### CODE REVIEW

File: storage/remote/queue_manager_test.go [hunk 1]
```diff
@@ -1441,6 +1470,110 @@ func TestIsSampleOld(t *testing.T) {
...
+
+	appendData(maxSamplesPerSend/2, timeShift, true)
+	time.Sleep(sampleAgeLimit)
+	appendData(maxSamplesPerSend/2, timeShift, true)
+	time.Sleep(sampleAgeLimit / 10)
+	appendData(maxSamplesPerSend/2, timeShift, true)
+	time.Sleep(2 * sampleAgeLimit)
+	appendData(2*maxSamplesPerSend, timeShift, false)
+	time.Sleep(sampleAgeLimit / 2)
+	c.SetStoreWait(0)
```
Comment (FUSAKLA) [inline storage/remote/queue_manager_test.go:1533]:
```suggestion
```

File: storage/remote/queue_manager_test.go [hunk 2]
```diff
@@ -1441,6 +1470,110 @@ func TestIsSampleOld(t *testing.T) {
...
+					Timestamp: s.T,
+					Value:     s.V,
+				})
+			}
+		}
+		batchID++
+	}
+	timeShift := -time.Millisecond * 5
+
+	c.SetStoreWait(0)
```
Comment (FUSAKLA) [inline storage/remote/queue_manager_test.go:1522]:
```suggestion
```

**Comment by david-vavra** [comment]
According to the failed test output, it seems that more _kinds_ of corruptions happens:
1. new series is created (probably by duplicating a sample and missing some of its labels)
```
                `{__name__="batch_3_id_12", batch_3_label_1="batch_3_label_1", batch_3_label_2="batch_3_label_2", batch_3_label_3="batch_3_label_3", batch_3_label_4="batch_3_label_4", batch_3_label_5="batch_3_label_5", batch_3_label_6="batch_3_label_6", batch_3_label_7="batch_3_label_7"}`:                                    {{Value: 9.131640722297644e+18, Timestamp: 1718268298528}},
        +       `{__name__="batch_3_id_12", batch_3_label_1="batch_3_label_1"}`:                                                                                                                                                                                                                                                      {{Value: 9.131640722297644e+18, Timestamp: 1718268298528}},

```
   - sometime it happens (once every few test runs) that new timeseries is created by mixing data from two separate batches (batch_3_id_4 metric getting some labels from batch2 which is supposed to get dropped because of sampleAgeLimit)
```
        +       `{__name__="batch_3_id_4", batch_2_label_3="batch_2_label_3", batch_2_label_4="batch_2_label_4", batch_3_label_1="batch_3_label_1", batch_3_label_2="batch_3_label_2"}`:                                                                       {{Value: 1.7243317006976353e+18, Timestamp: 1718269155316}},
                `{__name__="batch_3_id_4", batch_3_label_1="batch_3_label_1", batch_3_label_2="batch_3_label_2"}`:                                                                                                                                             {{Value: 1.7243317006976353e+18, Timestamp: 1718269155316}},
```
2. Timestamp and/or value of a given sample is changed
```
                `{__name__="batch_3_id_18", batch_3_label_1="batch_3_label_1", batch_3_label_2="batch_3_label_2", batch_3_label_3="batch_3_label_3", batch_3_label_4="batch_3_label_4", batch_3_label_5="batch_3_label_5", batch_3_label_6="batch_3_label_6", batch_3_label_7="batch_3_label_7", batch_3_label_8="batch_3_label_8"}`: {
                        {
        -                       Value:                1.8153395468049577e+18,
        +                       Value:                6.533624542662725e+17,
        -                       Timestamp:            1718268960809,
        +                       Timestamp:            1718268961310,
                                XXX_NoUnkeyedLiteral: {},
                                XXX_unrecognized:     nil,
                                XXX_sizecache:        0,
                        },
                },
```

**Comment by david-vavra** [comment]
I added the test case from this PR directly to https://github.com/prometheus/prometheus/pull/14078 . It will be easier to iterate on it within single PR together with the bug fix.

**Pull request closed without merging**